src/api_sketch.jl

```diff
@@ -0,0 +1,211 @@
+              to decode the function again. Examples incldue AtomGraph, WeaveMol (though
+              their current incarnations may need to be adapted)
+
+ENCODING: the process of translating a feature from its human-readable form (e.g. a 
+          float, string, etc.) to whatever will be ingested by ML model (could be as
+          simple as making a copy, more often is, e.g. building a one-hot vector)
+
+DECODING: The inverse process to encoding. Note that in many cases (e.g. a continuous-
+          valued feature encoded to a one-hot vector), the process isn't fully invertible,
+          i.e. you can't get back a precise value but rather only a range corresponding
+          to the associated onehot bin
+
+FEATURIZATION: the process of attaching encoded features to an atoms object (in a 
+               slight abuse of terminology, a featurization function  may also perform
+               feature encoding).
+
+File organization: probably eventually will be cleanest to have a `feat` folder with 
+separate files defining each feature type and similarly for the atoms object types and
+featurizations.
+
+In addition, there should be some place to drop functions for automatically building 
+features and featurizations for commonly desired features (i.e. things from lookup table
+and things making use of that dictionary of rdkit functions).
+
+For now, I'm sketching this all out in one file though.
+=#
+
+#= FEATURE OBJECTS
+
+I decided on an abstract type to allow the dispatch shown below in the "magical encoding"
+and "magical decoding" bits.
+
+Tn is the "natural" type of feature values, Te is the type of encoded values for the WHOLE
+STRUCTURE (not e.g. for a single atom).
+
+Representative examples – feature: feature object type, Tn, Te:
+
+block (s,p,d,f):            AtomFeat (contextual=false), String, Flux.OneHotMatrix 
+electronegativity (binned): AtomFeat (contextual=false), Float32, Flux.OneHotMatrix
+electronegativity (direct): AtomFeat (contextual=false), Float32, Vector{Float32}
+oxidation state:            AtomFeat (contextual=true) Int, Vector{Float32}
+distance between atoms:     PairFeat, Float32, Matrix{Float32}
+bond type:                  PairFeat, String, Array{Float32,3}
+
+(bond type is categorical and gets one-hot encoded, so the first two indices of the Matrix
+should be atom indices and the third should be indexing into the one-hot vector, which should just contain all zeros if the two atoms are not bonded (or we add a bin to the one-hot encoding to indicate that)
+
+All subtypes should define `encode_f` and `decode_f`
+`encode_f` should take in an atoms object and return Te
+`decode_f` should take in something of type Te and return something of type Tn
+=#
+abstract type Feature{Tn,Te} end
+
+# I think we should be able to do this too...there might be a more idiomatic way
+# to do what I'm showing below, perhaps via an "actual" holy trait type thing
```

Comment:
would love feedback on whether this is the best way to automate the dispatch I'm trying to do here